Never return nullptr silently

sjahr opened the pull request:
### Description

- Create some warning/ error message before returning a nullptr for better debugability
- Some minor documentation improvement

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/moveit/moveit2/blob/main/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://moveit.picknik.ai/humble/doc/examples/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/moveit/moveit2/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

sjahr commented on moveit_core/robot_state/src/robot_state.cpp:
```diff
@@ -911,24 +919,43 @@ const LinkModel* RobotState::getRigidlyConnectedParentLinkModel(const std::strin
     auto body{ getAttachedBody(object) };
     if (!body->hasSubframeTransform(frame))
+    {
+      // If the subframe transform is not found, log an error and return nullptr
```
```suggestion
```

rhaschke commented on moveit_core/robot_model/src/robot_model.cpp:
```diff
@@ -1357,21 +1357,21 @@ LinkModel* RobotModel::getLinkModel(const std::string& name, bool* has_link)
     return link;
+  }
   const moveit::core::LinkModel* parent_link = link->getParentLinkModel();
   const moveit::core::JointModel* joint = link->getParentJointModel();
```
A valid link always has a parent joint.

rhaschke commented on moveit_core/robot_model/src/robot_model.cpp:
```diff
@@ -1357,21 +1357,21 @@ LinkModel* RobotModel::getLinkModel(const std::string& name, bool* has_link)
   {
     *has_link = false;  // Report failure via argument
   }
-  else
-  {  // Otherwise print error
-    RCLCPP_ERROR(getLogger(), "Link '%s' not found in model '%s'", name.c_str(), model_name_.c_str());
-  }
```
The error message was printed only if the caller didn't provide the has_link argument to handle error checking itself. Probably, now you will get duplicates of the error message.

sjahr replied:
I deprecated the argument, this pattern is just ugly and prone to errors

rhaschke reviewed (commented):
Issueing an error message before returning a nullptr is not always desired: It might be perfectly possible that the caller correctly handles the nullptr result and issues an error itself or handles the error condition gracefully in some other fashion.
In that case, the error message would be spurious.
At most, you should issue a debug message.

sjahr commented on moveit_core/robot_model/src/robot_model.cpp:
```diff
@@ -1357,21 +1357,21 @@ LinkModel* RobotModel::getLinkModel(const std::string& name, bool* has_link)
   const moveit::core::JointModel* joint = link->getParentJointModel();
 
-  while (parent_link && joint->getType() == moveit::core::JointModel::FIXED)
+  while (parent_link && joint && joint->getType() == moveit::core::JointModel::FIXED)
```
```suggestion
  while (parent_link && joint->getType() == moveit::core::JointModel::FIXED)
```

marioprats commented:
I agree with Robert.
The functions returning `nullptr` are indicative of an error, which the caller should handle.
A better solution would be to introduce a good error propagation mechanism, like `tl::expected` or `absl::Status`. Then we could return an error with a string, and let the caller decide what to do.
In the meanwhile I think a log message makes sense, but I would make it a debug message.

rhaschke commented on moveit_core/robot_model/src/robot_model.cpp:
```diff
@@ -1331,7 +1331,8 @@ JointModel* RobotModel::getJointModel(const std::string& name)
 const LinkModel* RobotModel::getLinkModel(const std::string& name, bool* has_link) const
 {
-  return const_cast<RobotModel*>(this)->getLinkModel(name, has_link);
+  *has_link = false;
```
This will segfault if `has_link` is NULL!

rhaschke commented on moveit_core/robot_model/src/robot_model.cpp:
```diff
@@ -1347,27 +1348,29 @@ const LinkModel* RobotModel::getLinkModel(size_t index) const
-
-  if (has_link)
   {
-    *has_link = false;  // Report failure via argument
```
`has_link` was explicitly used to suppress the error message here and issue an error by the caller.

rhaschke commented on moveit_core/robot_model/src/robot_model.cpp:
```diff
@@ -1347,27 +1348,29 @@ const LinkModel* RobotModel::getLinkModel(size_t index) const
+const LinkModel* getLinkModel(const std::string& link, bool* has_link) const
+{
+  RCLCPP_ERROR(getLogger(), "'has_link' argument is deprecated. Use getLinkModel(const std::string& name) instead.");
+  *has_link = false;
```
segfault (see above)

rhaschke commented on moveit_core/robot_model/src/robot_model.cpp:
```diff
@@ -1347,27 +1348,29 @@ const LinkModel* RobotModel::getLinkModel(size_t index) const
 
+const LinkModel* getLinkModel(const std::string& link, bool* has_link) const
+{
+  RCLCPP_ERROR(getLogger(), "'has_link' argument is deprecated. Use getLinkModel(const std::string& name) instead.");
```
Issue a deprecation warning at compile-time only!

rhaschke reviewed (changes requested):
The last commit (57a85c88fd02728c18e24c282940cb7bce1ac823) should be handled in a separate PR. This is a fundamental change of the API and logic, which should be well motivated. I don't yet see the "ugliness" of the current API.
You should better address the reviews and change error messages to debug messages.

sjahr commented:
> The last commit ([57a85c8](https://github.com/moveit/moveit2/commit/57a85c88fd02728c18e24c282940cb7bce1ac823)) should be handled in a separate PR. This is a fundamental change of the API and logic, which should be well motivated. I don't yet see the "ugliness" of the current API. You should better address the reviews and change error messages to debug messages.

You're right, that derailed to initial intention of the PR too much, sorry. I reverted it and updated the log level to debug in most cases. Let me know what you think. 

>  I don't yet see the "ugliness" of the current API.

Maybe isn't the right term but rather outdated. I think we shouldn't pass raw boolean pointers anywhere in modern C++ but as you pointed out correctly, this PR is not the place to try changing that.

rhaschke commented on moveit_core/robot_model/src/robot_model.cpp:
```diff
@@ -1367,7 +1367,10 @@ LinkModel* RobotModel::getLinkModel(const std::string& name, bool* has_link)
 {
   if (!link)
+  {
+    RCLCPP_ERROR(getLogger(), "Cannot determine rigidly connected parent link because input link is nullptr");
```
The nullptr is correctly handled by callers. Change to debug level:
```suggestion
    RCLCPP_DEBUG(getLogger().get_child("getRigidlyConnectedParentLink", "Link is NULL");
```

rhaschke reviewed (commented):
See comments below.

rhaschke commented:
Sorry, I missed some closing parentheses.

sjahr commented on moveit_core/robot_model/src/robot_model.cpp:
```diff
@@ -1367,7 +1367,10 @@ LinkModel* RobotModel::getLinkModel(const std::string& name, bool* has_link)
 {
   if (!link)
+  {
+    RCLCPP_DEBUG(getLogger().get_child("getRigidlyConnectedParentLink", "Link is NULL");
```
```suggestion
    RCLCPP_DEBUG(getLogger().get_child("getRigidlyConnectedParentLink", "Link is NULL"));
```

sjahr commented on moveit_core/robot_state/src/robot_state.cpp:
```diff
@@ -911,24 +919,38 @@ const LinkModel* RobotState::getRigidlyConnectedParentLinkModel(const std::strin
     if (!body->hasSubframeTransform(frame))
+    {
+      RCLCPP_DEBUG(getLogger().get_child("getRigidlyConnectedParentLink",
+                   "Body '%s' does not have subframe '%s', object.c_str(), frame.c_str());
```
```suggestion
                   "Body '%s' does not have subframe '%s'", object.c_str(), frame.c_str());
```

rhaschke commented on moveit_core/robot_model/src/robot_model.cpp:
```diff
@@ -1367,7 +1367,10 @@ LinkModel* RobotModel::getLinkModel(const std::string& name, bool* has_link)
 {
   if (!link)
+  {
+    RCLCPP_DEBUG(getLogger().get_child("getRigidlyConnectedParentLink", "Link is NULL"));
```
```suggestion
    RCLCPP_DEBUG(getLogger().get_child("getRigidlyConnectedParentLink"), "Link is NULL");
```

rhaschke commented on moveit_core/robot_state/src/robot_state.cpp:
```diff
@@ -911,24 +919,38 @@ const LinkModel* RobotState::getRigidlyConnectedParentLinkModel(const std::strin
     std::string object{ frame.substr(0, idx) };
     if (!hasAttachedBody(object))
+    {
+      RCLCPP_DEBUG(getLogger().get_child("getRigidlyConnectedParentLink",
+                   "Attached object '%s' for frame '%s' does not exist.", object.c_str(), frame.c_str());
```
```suggestion
      RCLCPP_DEBUG(getLogger().get_child("getRigidlyConnectedParentLink"),
                   "Attached object '%s' for frame '%s' does not exist.", object.c_str(), frame.c_str());
```

rhaschke commented on moveit_core/robot_state/src/robot_state.cpp:
```diff
@@ -911,24 +919,38 @@ const LinkModel* RobotState::getRigidlyConnectedParentLinkModel(const std::strin
     auto body{ getAttachedBody(object) };
     if (!body->hasSubframeTransform(frame))
+    {
+      RCLCPP_DEBUG(getLogger().get_child("getRigidlyConnectedParentLink",
+                   "Body '%s' does not have subframe '%s'", object.c_str(), frame.c_str());
```
```suggestion
      RCLCPP_DEBUG(getLogger().get_child("getRigidlyConnectedParentLink"),
                   "Body '%s' does not have subframe '%s'", object.c_str(), frame.c_str());
```

rhaschke reviewed (commented):
Fix parentheses. Still unchecked. I just did that in the Web GUI.

henningkayser commented:
@sjahr unfortunately this needs a new iteration because I merged #2985
